# Nearest Petrol Stations Bot - DATA part

Goal of this project is to help drivers to navigate the nearest petrol stations in all cities and regions of Kazakhstan, but first bot should be able do navigate in two major cities of Kazakhstan - Astana and Almaty. But it is not going to find petrol stations solely by distance, but also on drivers' preferences of petrol station organization (e.g. Helios, QazOil, Gazprom), fuel (gasoline, gas, diesel) and type of this fuel (95, 98, 92, etc). Furthermore, it will only show these petrol stations that have at least 30% of tank filled with specific fuel chosen by user.
    
    In order to fullfil this objective, we need to clean and prepare data. "Almaty.xlsx" & "Astana.xlsx" will act as main dictionaries that have names, physical addresses and coordinates (Langitude, Latitude) of petrol stations in Almaty and Astana. 
    
    (It is a proccess of creating a local bot - connecting bot to a database is not written). 

## Importing libraries

In [2]:
import pandas as pd # managing dataframes.
import numpy as np  
import math # math & numpy are for numerical/mathematical operations. 
import requests # needed to extract coordinates from physical addresses of petrol stations that lack coordinates. 

## Preparing dictionary and finding null-values of Almaty city dataset

In [38]:
SUNP_Almaty_null = pd.read_excel('Алматы.xlsx')
SUNP_Almaty_null.info()
SUNP_Almaty_null.head(2)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 286 entries, 0 to 285
Data columns (total 8 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   NameRu      286 non-null    object 
 1   Bin         285 non-null    float64
 2   Address     280 non-null    object 
 3   Latitude    84 non-null     object 
 4   Longitude   84 non-null     object 
 5   ShortName   261 non-null    object 
 6   Unnamed: 6  10 non-null     float64
 7   Unnamed: 7  1 non-null      float64
dtypes: float64(3), object(5)
memory usage: 18.0+ KB


NameRu           Bin  \
0                ТОО "Кулнафта"  1.010400e+11   
1  АЗС №14 Алматы ТОО "Sinooil"  9.054102e+10   

                                             Address   Latitude  Longitude  \
0            г. Алматы, проспект Аль- Фараби, дом 17  43.230017  76.945627   
1  Суюнбая, 150 в                                ...   43.33054  76.959311   

                      ShortName  Unnamed: 6  Unnamed: 7  
0                ТОО "Кулнафта"         NaN         NaN  
1  АЗС №14 Алматы ТОО "Sinooil"         NaN         NaN

First, I am removing `Bin`, `Unnamed: 6`, and `Unnamed: 7` as these columns are not important for the analysis.

In [39]:
SUNP_Almaty_null = SUNP_Almaty_null.drop(['Bin', 'Unnamed: 6', 'Unnamed: 7'], axis = 1)
SUNP_Almaty_null = SUNP_Almaty_null.rename(columns = {'NameRu':'AZS'})

`Almaty` contents: 
 - `AZS` - Name of the petrol station
 - `Address` - Physical address
 - `Latitude` - lattiude coordinate
 - `Langitude` - langitude coordinate
 - `ShortName` - shortname of the petrol station

In `Almaty` dataset, we have 6 null values in Address and 202 null values in Latitude and Longitude columns. While it is easy to extract coordinates from physical addresses, it is better to clear out petrol stations that do not have their physical addresses written.

In [40]:
null_AZS = SUNP_Almaty_null[SUNP_Almaty_null['Address'].isna()]
null_AZS
# These are the names of petrol stations that do not have physicall addresses written.

AZS Address   Latitude  Longitude  \
4                  НБ "Синтез"     NaN        NaN        NaN   
84   АЗС ELF ИП "Буханцов А.А"     NaN        NaN        NaN   
203           АЗС №102 "SOCAR"     NaN        NaN        NaN   
205           АЗС №110 "SOCAR"     NaN  43.226205  76.844562   
206           АЗС №113 "SOCAR"     NaN        NaN        NaN   
228           АЗС №122 "SOCAR"     NaN        NaN        NaN   

                     ShortName  
4                  НБ "Синтез"  
84   АЗС ELF ИП "Буханцов А.А"  
203           АЗС №102 "SOCAR"  
205           АЗС №110 "SOCAR"  
206           АЗС №113 "SOCAR"  
228           АЗС №122 "SOCAR"

In [41]:
SUNP_Almaty_null = SUNP_Almaty_null.dropna(subset = ['Address'])
SUNP_Almaty_null.info()
SUNP_Almaty_null.head()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 280 entries, 0 to 285
Data columns (total 5 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   AZS        280 non-null    object
 1   Address    280 non-null    object
 2   Latitude   83 non-null     object
 3   Longitude  83 non-null     object
 4   ShortName  255 non-null    object
dtypes: object(5)
memory usage: 13.1+ KB


AZS  \
0                ТОО "Кулнафта"   
1  АЗС №14 Алматы ТОО "Sinooil"   
2    АЗС №50 ТОО "Ройал Петрол"   
3    АЗС №70 ТОО "Ройал Петрол"   
5                 АЗС "CASTROL"   

                                             Address   Latitude  Longitude  \
0            г. Алматы, проспект Аль- Фараби, дом 17  43.230017  76.945627   
1  Суюнбая, 150 в                                ...   43.33054  76.959311   
2                             мкр. Каменька дом 26/1        NaN        NaN   
3                                    Розыбакиева 262  43.216184  76.893288   
5                                  ул. Майлина, 79 Б        NaN        NaN   

                      ShortName  
0                ТОО "Кулнафта"  
1  АЗС №14 Алматы ТОО "Sinooil"  
2    АЗС №50 ТОО "Ройал Петрол"  
3    АЗС №70 ТОО "Ройал Петрол"  
5                   АЗС CASTROL

Now there are 280 rows and 280 non-null values in `Address` section. It can be observed that there are only 83 petrol stations that have both coordinates (Latitude, Langitude) and physical addresses. That's why I need to prepare the physical addresses of petrol stations that lack coordinates and use them in `GOOGLE API` to extract coordinates.

In [42]:
def fix_address_almaty(row): # function that adds country + city name to the Address section, as not all addresses have it.
    return "Казахстан, г.Алматы, " + row
SUNP_Almaty_null['Address'] = SUNP_Almaty_null['Address'].apply(fix_address_almaty)
SUNP_Almaty_null.head()

AZS  \
0                ТОО "Кулнафта"   
1  АЗС №14 Алматы ТОО "Sinooil"   
2    АЗС №50 ТОО "Ройал Петрол"   
3    АЗС №70 ТОО "Ройал Петрол"   
5                 АЗС "CASTROL"   

                                             Address   Latitude  Longitude  \
0  Казахстан, г.Алматы, г. Алматы, проспект Аль- ...  43.230017  76.945627   
1  Казахстан, г.Алматы, Суюнбая, 150 в           ...   43.33054  76.959311   
2        Казахстан, г.Алматы, мкр. Каменька дом 26/1        NaN        NaN   
3               Казахстан, г.Алматы, Розыбакиева 262  43.216184  76.893288   
5             Казахстан, г.Алматы, ул. Майлина, 79 Б        NaN        NaN   

                      ShortName  
0                ТОО "Кулнафта"  
1  АЗС №14 Алматы ТОО "Sinooil"  
2    АЗС №50 ТОО "Ройал Петрол"  
3    АЗС №70 ТОО "Ройал Петрол"  
5                   АЗС CASTROL

After using `fix_address_almaty` function, some of the addresses are not written correctly as they already had country and city names. That's why I am going to fix it.

In [44]:
SUNP_Almaty_null['Address'] = SUNP_Almaty_null['Address'].str.replace('Казахстан, г.Алматы, Казахстан, г. Алматы', 'Казахстан, г.Алматы')
SUNP_Almaty_null['Address'] = SUNP_Almaty_null['Address'].str.replace('г.Алматы, г. Алматы', 'г.Алматы')
filtered_rows = SUNP_Almaty_null[SUNP_Almaty_null['Address'].str.contains('Казахстан, г.Алматы, г. Алматы')]
length = len(filtered_rows)
print('Number of addresses with duplicated names of country and city:', length, '\n')
SUNP_Almaty_null 

Number of addresses with duplicated names of country and city: 0 



C:\Users\Beibarys Nyussupov\AppData\Local\Temp\ipykernel_21360\3856561085.py:1: FutureWarning: The default value of regex will change from True to False in a future version.
  SUNP_Almaty_null['Address'] = SUNP_Almaty_null['Address'].str.replace('Казахстан, г.Алматы, Казахстан, г. Алматы', 'Казахстан, г.Алматы')
C:\Users\Beibarys Nyussupov\AppData\Local\Temp\ipykernel_21360\3856561085.py:2: FutureWarning: The default value of regex will change from True to False in a future version.
  SUNP_Almaty_null['Address'] = SUNP_Almaty_null['Address'].str.replace('г.Алматы, г. Алматы', 'г.Алматы')


AZS  \
0                                       ТОО "Кулнафта"   
1                         АЗС №14 Алматы ТОО "Sinooil"   
2                           АЗС №50 ТОО "Ройал Петрол"   
3                           АЗС №70 ТОО "Ройал Петрол"   
5                                        АЗС "CASTROL"   
..                                                 ...   
281                                   АЗС №129 "SOCAR"   
282  АЗС 156 ТОО "Газпромнефть-Казахствн Операционн...   
283  АЗС 157 ТОО "Газпромнефть-Казахствн Операционн...   
284                              Азс Гелиос Алматы №89   
285                                   АЗС №101 "SOCAR"   

                                               Address   Latitude  Longitude  \
0    Казахстан, г.Алматы, проспект Аль- Фараби, дом 17  43.230017  76.945627   
1    Казахстан, г.Алматы, Суюнбая, 150 в           ...   43.33054  76.959311   
2          Казахстан, г.Алматы, мкр. Каменька дом 26/1        NaN        NaN   
3                 Казахстан, г.Алматы, Розыбакиева 262  43.216184  76.893288   
5               Казахстан, г.Алматы, ул. Майлина, 79 Б        NaN        NaN   
..                                                 ...        ...        ...   
281            Казахстан, г.Алматы, ул.Бокейханова 241        NaN        NaN   
282            Казахстан, г.Алматы, ул. Ержанова 135/1        NaN        NaN   
283        Казахстан, г.Алматы, мкр. Казахфильм 34 "Б"        NaN        NaN   
284  Казахстан, г.Алматы, мкр Акжар, ул. Даулеткере...        NaN        NaN   
285            Казахстан, г.Алматы, мкр. Болашак д.1/7        NaN        NaN   

                        ShortName  
0                  ТОО "Кулнафта"  
1    АЗС №14 Алматы ТОО "Sinooil"  
2      АЗС №50 ТОО "Ройал Петрол"  
3      АЗС №70 ТОО "Ройал Петрол"  
5                     АЗС CASTROL  
..                            ...  
281                           NaN  
282                           NaN  
283                           NaN  
284                           NaN  
285                           NaN  

[280 rows x 5 columns]

In [45]:
SUNP_Almaty_null['Address'].value_counts() # Looks correct! 

Казахстан, г.Алматы, ул. Тлендиева 84                                   2
Казахстан, г.Алматы, пр.Суюнбая 66/6                                    2
Казахстан, г.Алматы, пр.Райымбека 50                                    2
Казахстан, г.Алматы, проспект Аль- Фараби, дом 17                       1
Казахстан, г.Алматы, улица Булкышева, 9 "в"                             1
                                                                       ..
Казахстан, г.Алматы, Ауэзовский мкр. Аксай-3 "Б" дом 18-Б               1
Казахстан, г.Алматы, Алатауский район мкр. Акбулак АЗТМ участок №285    1
Казахстан, г.Алматы, Турксибский район мкр. Колхозшы дом 6/3            1
Казахстан, г.Алматы, Жетысуйский район пер. Первомайский д. 35/2        1
Казахстан, г.Алматы, мкр. Болашак д.1/7                                 1
Name: Address, Length: 277, dtype: int64

It looks correct and the error is fixed. Now I will do the same proccess with `Astana` city dataset.

In [47]:
SUNP_Astana_null = pd.read_excel('Астана.xlsx')
SUNP_Astana_null.info()
SUNP_Astana_null.head(1)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 145 entries, 0 to 144
Data columns (total 21 columns):
 #   Column       Non-Null Count  Dtype         
---  ------       --------------  -----         
 0   NameRu       145 non-null    object        
 1   Bin          145 non-null    int64         
 2   Address      145 non-null    object        
 3   Latitude     66 non-null     float64       
 4   Longitude    66 non-null     float64       
 5   ShortName    133 non-null    object        
 6   Unnamed: 6   8 non-null      object        
 7   Unnamed: 7   1 non-null      object        
 8   Unnamed: 8   0 non-null      float64       
 9   Unnamed: 9   1 non-null      float64       
 10  Unnamed: 10  1 non-null      float64       
 11  Unnamed: 11  1 non-null      datetime64[ns]
 12  Unnamed: 12  1 non-null      object        
 13  Unnamed: 13  1 non-null      float64       
 14  Unnamed: 14  1 non-null      float64       
 15  Unnamed: 15  1 non-null      float64       
 16  Unnamed:

NameRu           Bin            Address   Latitude  Longitude  \
0  ТОО "МНПЗ "AMAL"  161140000255  ул. Ондирис, 60/1  51.261417  71.368687   

    ShortName Unnamed: 6 Unnamed: 7  Unnamed: 8  Unnamed: 9  ...  Unnamed: 11  \
0  ТОО "Амал"        NaN        NaN         NaN         NaN  ...          NaT   

  Unnamed: 12 Unnamed: 13  Unnamed: 14  Unnamed: 15  Unnamed: 16  Unnamed: 17  \
0         NaN         NaN          NaN          NaN          NaN          NaN   

   Unnamed: 18  Unnamed: 19  Unnamed: 20  
0          NaN          NaN          NaN  

[1 rows x 21 columns]

`Astana` dataset has useless `Bin`, and `Unnamed` columns that I am going to filter out from dataframe.

In [48]:
SUNP_Astana_null = SUNP_Astana_null.drop(columns = SUNP_Astana_null.columns[6:22], axis = 1)
SUNP_Astana_null = SUNP_Astana_null.drop(columns = ['Bin'])
SUNP_Astana_null = SUNP_Astana_null.rename(columns = {'NameRu':'AZS'})

In [51]:
SUNP_Astana_null = SUNP_Astana_null.dropna(subset = ['Address'])
SUNP_Astana_null.info()
SUNP_Astana_null.head()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 145 entries, 0 to 144
Data columns (total 5 columns):
 #   Column     Non-Null Count  Dtype  
---  ------     --------------  -----  
 0   AZS        145 non-null    object 
 1   Address    145 non-null    object 
 2   Latitude   66 non-null     float64
 3   Longitude  66 non-null     float64
 4   ShortName  133 non-null    object 
dtypes: float64(2), object(3)
memory usage: 5.8+ KB


AZS                       Address   Latitude  Longitude  \
0  ТОО "МНПЗ "AMAL"             ул. Ондирис, 60/1  51.261417  71.368687   
1    АЗС Аурика №16        ш. Астана-Караганда 29        NaN        NaN   
2          АЗС Амал        Шоссе Ондирис, д. 60/1  51.246101  71.379093   
3     АЗС №12 "Ойл"  ул. Тауелсыздык проспект, 4а  51.150613  71.457811   
4     АЗС Аурика №3            ул. Жирентаева 6/1  51.148455  71.475729   

       ShortName  
0     ТОО "Амал"  
1  АЗС Аурика 16  
2       АЗС Амал  
3  АЗС №12 "Ойл"  
4  АЗС Аурика №3

All petrol stations in `Astana` dataframe have addresses written, but only 66 of petrol stations have both physical addresses and coordinates (Langitude & Latitude), other 79 petrol stations lack coordinates. Again, I am going to prepare physical addresses of this dataset for using them in `GOOGLE API` and clear the errors that will be created in further function.

In [52]:
def fix_address_astana(row): # function that adds names country (Kazakhstan) + city (Astana) to address section of petrol stations.
    return "Казахстан, г.Астана, " + row
SUNP_Astana_null['Address'] = SUNP_Astana_null['Address'].apply(fix_address_astana)
SUNP_Astana_null['Address'] = SUNP_Astana_null['Address'].str.replace('Казахстан, г.Астана, Казахстан, г. Астана', 'Казахстан, г.Астана')
SUNP_Astana_null['Address'] = SUNP_Astana_null['Address'].str.replace('г.Астана, г. Астана', 'г.Астана')
filtered_rows = SUNP_Astana_null[SUNP_Astana_null['Address'].str.contains('Казахстан, г.Астана, г. Астана')]
length = len(filtered_rows)
print('Number of addresses with duplicated country and city names:', length, '\n')
SUNP_Astana_null['Address'].value_counts().head()

0 



C:\Users\Beibarys Nyussupov\AppData\Local\Temp\ipykernel_21360\134380619.py:4: FutureWarning: The default value of regex will change from True to False in a future version.
  SUNP_Astana_null['Address'] = SUNP_Astana_null['Address'].str.replace('Казахстан, г.Астана, Казахстан, г. Астана', 'Казахстан, г.Астана')
C:\Users\Beibarys Nyussupov\AppData\Local\Temp\ipykernel_21360\134380619.py:5: FutureWarning: The default value of regex will change from True to False in a future version.
  SUNP_Astana_null['Address'] = SUNP_Astana_null['Address'].str.replace('г.Астана, г. Астана', 'г.Астана')


Казахстан, г.Астана, ул. Ондирис, 60/1            1
Казахстан, г.Астана, ш. Алаш 10/2                 1
Казахстан, г.Астана, Кургальджинское шоссе  56    1
Казахстан, г.Астана, улица С 622 д 5              1
Казахстан, г.Астана, ул.Айнаколь 62/2             1
Name: Address, dtype: int64

In [12]:
SUNP_Astana_null.head()

AZS                                            Address  \
0  ТОО "МНПЗ "AMAL"             Казахстан, г.Астана, ул. Ондирис, 60/1   
1    АЗС Аурика №16        Казахстан, г.Астана, ш. Астана-Караганда 29   
2          АЗС Амал        Казахстан, г.Астана, Шоссе Ондирис, д. 60/1   
3     АЗС №12 "Ойл"  Казахстан, г.Астана, ул. Тауелсыздык проспект, 4а   
4     АЗС Аурика №3            Казахстан, г.Астана, ул. Жирентаева 6/1   

    Latitude  Longitude      ShortName  
0  51.261417  71.368687     ТОО "Амал"  
1        NaN        NaN  АЗС Аурика 16  
2  51.246101  71.379093       АЗС Амал  
3  51.150613  71.457811  АЗС №12 "Ойл"  
4  51.148455  71.475729  АЗС Аурика №3

## Getting coordinates for petrol stations 
I am joining `SUNP_Almaty_null` and `SUNP_Astana_null` dataframes together so it will be easier to process data and extract coordinates at further steps.

In [20]:
main_data_merged_null = pd.concat([SUNP_Astana_null, SUNP_Almaty_null], ignore_index=True)
main_data_merged_null.info()
main_data_merged_null

NameError: name 'SUNP_Astana_null' is not defined

Next step is to clean coordinates columns (Latitude & Langitude). Here is what I am going to do:
 - Dropping rows that have NA values in 3 columns at the same time (Address & Latitude & Longitude)
 - Dropping rows that have `qwe` written in Latitude and Longitude columns. 
 - Transforming `Longitude and Latitude` columns into float-type columns.

In [54]:
# Dropping rows where all three columns are NA
main_data_merged_null = main_data_merged_null = main_data_merged_null.dropna(subset = ['Latitude', 'Longitude', 'Address'], how = 'all')

# Dropping rows with 'qwe' string in Latitude and longitude. 
main_data_merged_null= main_data_merged_null[main_data_merged_null['Latitude'] != 'qwe']

# Transforming Latitude and Longitude columns into float-type columns.
main_data_merged_null['Latitude'] = main_data_merged_null['Latitude'].astype(float)
main_data_merged_null['Longitude'] = main_data_merged_null['Longitude'].astype(float)
main_data_merged_null.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 424 entries, 0 to 424
Data columns (total 5 columns):
 #   Column     Non-Null Count  Dtype  
---  ------     --------------  -----  
 0   AZS        424 non-null    object 
 1   Address    424 non-null    object 
 2   Latitude   148 non-null    float64
 3   Longitude  148 non-null    float64
 4   ShortName  387 non-null    object 
dtypes: float64(2), object(3)
memory usage: 19.9+ KB


The next step that should be done is to create a new dataframe that will have all rows with no coordinates. I will apply `GOOGLE API` on these rows and extract coordinates.
 - `url`:link from where the coordinates will be extracted
 - `indicator1`: filter for main dataframe where `Latitude` is NA
 - `indicator2`: filter for main dataframe where `Longitude` is NA
 - `df`: new dataframe created by filtering rows that don't have coordinates from main dataframe `main_data_merged_null`

In [55]:
url = "https://maps.googleapis.com/maps/api/geocode/json?"
indicator = main_data_merged_null["Latitude"].isna()
indicator1 = main_data_merged_null["Longitude"].isna()
df = main_data_merged_null[indicator & indicator1]
df

AZS  \
1                                       АЗС Аурика №16   
6                                       АЗС Аурика №12   
10                                      АЗС Аурика №19   
13                                      АЗС Аурика №15   
16                                        АЗС №7 "Ойл"   
..                                                 ...   
420                                   АЗС №129 "SOCAR"   
421  АЗС 156 ТОО "Газпромнефть-Казахствн Операционн...   
422  АЗС 157 ТОО "Газпромнефть-Казахствн Операционн...   
423                              Азс Гелиос Алматы №89   
424                                   АЗС №101 "SOCAR"   

                                               Address  Latitude  Longitude  \
1          Казахстан, г.Астана, ш. Астана-Караганда 29       NaN        NaN   
6                      Казахстан, г.Астана, ш. Алаш 73       NaN        NaN   
10                    Казахстан, г.Астана, ул.187 д.28       NaN        NaN   
13          Казахстан, г.Астана, ул. Аль Фараби д 24/1       NaN        NaN   
16          Казахстан, г.Астана, ул. Кудайбердыйлы 19B       NaN        NaN   
..                                                 ...       ...        ...   
420            Казахстан, г.Алматы, ул.Бокейханова 241       NaN        NaN   
421            Казахстан, г.Алматы, ул. Ержанова 135/1       NaN        NaN   
422        Казахстан, г.Алматы, мкр. Казахфильм 34 "Б"       NaN        NaN   
423  Казахстан, г.Алматы, мкр Акжар, ул. Даулеткере...       NaN        NaN   
424            Казахстан, г.Алматы, мкр. Болашак д.1/7       NaN        NaN   

          ShortName  
1     АЗС Аурика 16  
6        Аурика №12  
10    АЗС Аурика 19  
13   АЗС Аурика №15  
16     АЗС №7 "Ойл"  
..              ...  
420             NaN  
421             NaN  
422             NaN  
423             NaN  
424             NaN  

[276 rows x 5 columns]

276 rows that do not have coordinates in a total. I've prepared a function `get_coords` that will extract coordinates from physical addresses below:

In [56]:
# Function that extracts coordinates and places them in a form of (lat, lon) in lat/lon column.
def get_coords(x):
    response = requests.get('https://maps.googleapis.com/maps/api/geocode/json?address={}&key=AIzaSyCKukxFH4qb2nip9B5nK1eUQB8l_btWqBU'.format(x))
    resp_json_payload = response.json()
    
    if resp_json_payload['results']:
       lat = resp_json_payload['results'][0]['geometry']['location']['lat']
       long= resp_json_payload['results'][0]['geometry']['location']['lng']
    else:
        lat = None
        long = None
    
    return lat, long

df['lat/lon'] = df['Address'].apply(lambda x: get_coords(x))
df

C:\Users\Beibarys Nyussupov\AppData\Local\Temp\ipykernel_21360\2716411875.py:15: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df['lat/lon'] = df['Address'].apply(lambda x: get_coords(x))


AZS  \
1                                       АЗС Аурика №16   
6                                       АЗС Аурика №12   
10                                      АЗС Аурика №19   
13                                      АЗС Аурика №15   
16                                        АЗС №7 "Ойл"   
..                                                 ...   
420                                   АЗС №129 "SOCAR"   
421  АЗС 156 ТОО "Газпромнефть-Казахствн Операционн...   
422  АЗС 157 ТОО "Газпромнефть-Казахствн Операционн...   
423                              Азс Гелиос Алматы №89   
424                                   АЗС №101 "SOCAR"   

                                               Address  Latitude  Longitude  \
1          Казахстан, г.Астана, ш. Астана-Караганда 29       NaN        NaN   
6                      Казахстан, г.Астана, ш. Алаш 73       NaN        NaN   
10                    Казахстан, г.Астана, ул.187 д.28       NaN        NaN   
13          Казахстан, г.Астана, ул. Аль Фараби д 24/1       NaN        NaN   
16          Казахстан, г.Астана, ул. Кудайбердыйлы 19B       NaN        NaN   
..                                                 ...       ...        ...   
420            Казахстан, г.Алматы, ул.Бокейханова 241       NaN        NaN   
421            Казахстан, г.Алматы, ул. Ержанова 135/1       NaN        NaN   
422        Казахстан, г.Алматы, мкр. Казахфильм 34 "Б"       NaN        NaN   
423  Казахстан, г.Алматы, мкр Акжар, ул. Даулеткере...       NaN        NaN   
424            Казахстан, г.Алматы, мкр. Болашак д.1/7       NaN        NaN   

          ShortName                         lat/lon  
1     АЗС Аурика 16        (49.8046835, 73.1093826)  
6        Аурика №12        (51.2219808, 71.5267412)  
10    АЗС Аурика 19        (51.1751698, 71.3821404)  
13   АЗС Аурика №15         (51.183359, 71.4514793)  
16     АЗС №7 "Ойл"        (51.1655126, 71.4272222)  
..              ...                             ...  
420             NaN        (43.2996335, 76.8985325)  
421             NaN        (43.3339054, 76.9378297)  
422             NaN          (43.19648, 76.9059625)  
423             NaN  (43.181349, 76.80579639999999)  
424             NaN         (43.285475, 76.9995418)  

[276 rows x 6 columns]

Let's check whether there are rows that failed to generate coordinates.

In [58]:
null_indicator = df['lat/lon'].isna()
len(df[df['lat/lon'] == null_indicator]) # There are not coordinates that failed to generate coordinates.

0

Now I need to remove brackets from values in `lat/lon` column, so there will be only ' ' (spacebar) between Latitude and Longitude. Then, I am diving these coordinates into `Latitude` and `Longitude` columns.

In [59]:
# Removing brackets and old coordinates (Latitude & Longitude) columns
df['lat/lon'] = df['lat/lon'].astype(str).str.replace('(', '').str.replace(')', '').str.replace(',', '')
columns_exclude = ['Latitude', 'Longitude']
df.drop(columns = columns_exclude, inplace = True)

# Splitting coordinates in lat/lon column into two Latitude and Longitude columns
df[['Latitude', 'Longitude']] = df['lat/lon'].str.split(' ', expand=True)

# Converting columns with coordinates to float type columns
df['Latitude'] = df['Latitude'].astype(float)
df['Longitude'] = df['Longitude'].astype(float)

df.info()
df.head()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 276 entries, 1 to 424
Data columns (total 6 columns):
 #   Column     Non-Null Count  Dtype  
---  ------     --------------  -----  
 0   AZS        276 non-null    object 
 1   Address    276 non-null    object 
 2   ShortName  239 non-null    object 
 3   lat/lon    276 non-null    object 
 4   Latitude   276 non-null    float64
 5   Longitude  276 non-null    float64
dtypes: float64(2), object(4)
memory usage: 15.1+ KB


C:\Users\Beibarys Nyussupov\AppData\Local\Temp\ipykernel_21360\2113401583.py:2: FutureWarning: The default value of regex will change from True to False in a future version. In addition, single character regular expressions will *not* be treated as literal strings when regex=True.
  df['lat/lon'] = df['lat/lon'].astype(str).str.replace('(', '').str.replace(')', '').str.replace(',', '')
C:\Users\Beibarys Nyussupov\AppData\Local\Temp\ipykernel_21360\2113401583.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df['lat/lon'] = df['lat/lon'].astype(str).str.replace('(', '').str.replace(')', '').str.replace(',', '')
C:\Users\Beibarys Nyussupov\AppData\Local\Temp\ipykernel_21360\2113401583.py:4: SettingWithCopyWarning: 
A value is trying to be set on 

Now there are no NA values in Address, Latitude or Longitude columns. Let's merge df with our main dataframe.

In [61]:
# Joining df with ain dataframe
main_data_merged_null_final = pd.concat([main_data_merged_null, df], ignore_index = True, axis = 0)

# Checking for possible NA values and dropping lat/lon column as it is not  neaded in the final dataframe
main_data_merged_null_final = main_data_merged_null_final.dropna(subset = ['Latitude', 'Longitude'], how = 'all')
main_data_merged_null_final = main_data_merged_null_final.drop(columns = ['lat/lon'])


main_data_merged_null_final.info()
main_data_merged_null_final.head()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 424 entries, 0 to 699
Data columns (total 5 columns):
 #   Column     Non-Null Count  Dtype  
---  ------     --------------  -----  
 0   AZS        424 non-null    object 
 1   Address    424 non-null    object 
 2   Latitude   424 non-null    float64
 3   Longitude  424 non-null    float64
 4   ShortName  387 non-null    object 
dtypes: float64(2), object(3)
memory usage: 19.9+ KB


AZS                                            Address  \
0  ТОО "МНПЗ "AMAL"             Казахстан, г.Астана, ул. Ондирис, 60/1   
2          АЗС Амал        Казахстан, г.Астана, Шоссе Ондирис, д. 60/1   
3     АЗС №12 "Ойл"  Казахстан, г.Астана, ул. Тауелсыздык проспект, 4а   
4     АЗС Аурика №3            Казахстан, г.Астана, ул. Жирентаева 6/1   
5     АЗС Аурика №1                  Казахстан, г.Астана, пр. Туран 95   

    Latitude  Longitude      ShortName  
0  51.261417  71.368687     ТОО "Амал"  
2  51.246101  71.379093       АЗС Амал  
3  51.150613  71.457811  АЗС №12 "Ойл"  
4  51.148455  71.475729  АЗС Аурика №3  
5  51.073720  71.388683      Аурика №1

Geojsons are not necessary for this project, but I've added a function to generate geojsons to the main datframe as `geojson_region` just in case if you need it.

In [127]:
main_data_merged_null_final['geojson_region'] = ''
def get_circle_coord(theta, x_center, y_center, radius):
    x = radius * math.cos(theta) + x_center
    y = radius * math.sin(theta) + y_center
    return [y,x]

# Эта функция возвращает все координаты n-угольника, в зависимости от указанного радиуса, широты и долготы, и кол-ва точек
def get_all_circle_coords(x_center, y_center, radius, n_points):
    # переводим радиус из метров для работы с широтой долготой
    radius = radius * 90 / 10000000 
    thetas = [i/n_points * math.tau for i in range(n_points)]
    circle_coords = [get_circle_coord(theta, x_center, y_center, radius) for theta in thetas]
    circle_coords.append(get_circle_coord(0, x_center, y_center, radius))
    return circle_coords

for i, row in main_data_merged_null_final.iterrows():
    x_center = row['Latitude']
    y_center = row['Longitude']
    radius = 50
    n_points = 1000
    circle_coords = get_all_circle_coords(x_center, y_center, radius, n_points)
    main_data_merged_null_final['geojson_region'][i] = '{\"type\": \"FeatureCollection\", \"crs\": {\"type\": \"name\", \"properties\": {\"name\": \"urn:ogc:def:crs:OGC:1.3:CRS84\"}}, \"features\":' + '['+'{\"type\": \"Feature\", \"properties\": {\"ISO\": 15, \"NAME_1\": \"test\"}, \"geometry\": {\"type\": \"Polygon\", \"coordinates\":' + '[' + '{}'.format(circle_coords) + ']}}' + ']}'
    
main_data_merged_null_final.info()
main_data_merged_null_final

C:\Users\Beibarys Nyussupov\AppData\Local\Temp\ipykernel_41116\924617788.py:22: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  main_data_merged_null_final['geojson_region'][i] = '{\"type\": \"FeatureCollection\", \"crs\": {\"type\": \"name\", \"properties\": {\"name\": \"urn:ogc:def:crs:OGC:1.3:CRS84\"}}, \"features\":' + '['+'{\"type\": \"Feature\", \"properties\": {\"ISO\": 15, \"NAME_1\": \"test\"}, \"geometry\": {\"type\": \"Polygon\", \"coordinates\":' + '[' + '{}'.format(circle_coords) + ']}}' + ']}'


<class 'pandas.core.frame.DataFrame'>
Int64Index: 424 entries, 0 to 699
Data columns (total 6 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   AZS             424 non-null    object 
 1   Address         424 non-null    object 
 2   Latitude        424 non-null    float64
 3   Longitude       424 non-null    float64
 4   ShortName       387 non-null    object 
 5   geojson_region  424 non-null    object 
dtypes: float64(2), object(4)
memory usage: 39.4+ KB


AZS  \
0                                     ТОО "МНПЗ "AMAL"   
2                                             АЗС Амал   
3                                        АЗС №12 "Ойл"   
4                                        АЗС Аурика №3   
5                                        АЗС Аурика №1   
..                                                 ...   
695                                   АЗС №129 "SOCAR"   
696  АЗС 156 ТОО "Газпромнефть-Казахствн Операционн...   
697  АЗС 157 ТОО "Газпромнефть-Казахствн Операционн...   
698                              Азс Гелиос Алматы №89   
699                                   АЗС №101 "SOCAR"   

                                               Address   Latitude  Longitude  \
0               Казахстан, г.Астана, ул. Ондирис, 60/1  51.261417  71.368687   
2          Казахстан, г.Астана, Шоссе Ондирис, д. 60/1  51.246101  71.379093   
3    Казахстан, г.Астана, ул. Тауелсыздык проспект, 4а  51.150613  71.457811   
4              Казахстан, г.Астана, ул. Жирентаева 6/1  51.148455  71.475729   
5                    Казахстан, г.Астана, пр. Туран 95  51.073720  71.388683   
..                                                 ...        ...        ...   
695            Казахстан, г.Алматы, ул.Бокейханова 241  43.299633  76.898533   
696            Казахстан, г.Алматы, ул. Ержанова 135/1  43.333905  76.937830   
697        Казахстан, г.Алматы, мкр. Казахфильм 34 "Б"  43.196480  76.905963   
698  Казахстан, г.Алматы, мкр Акжар, ул. Даулеткере...  43.181349  76.805796   
699            Казахстан, г.Алматы, мкр. Болашак д.1/7  43.285475  76.999542   

         ShortName                                     geojson_region  
0       ТОО "Амал"  {"type": "FeatureCollection", "crs": {"type": ...  
2         АЗС Амал  {"type": "FeatureCollection", "crs": {"type": ...  
3    АЗС №12 "Ойл"  {"type": "FeatureCollection", "crs": {"type": ...  
4    АЗС Аурика №3  {"type": "FeatureCollection", "crs": {"type": ...  
5        Аурика №1  {"type": "FeatureCollection", "crs": {"type": ...  
..             ...                                                ...  
695            NaN  {"type": "FeatureCollection", "crs": {"type": ...  
696            NaN  {"type": "FeatureCollection", "crs": {"type": ...  
697            NaN  {"type": "FeatureCollection", "crs": {"type": ...  
698            NaN  {"type": "FeatureCollection", "crs": {"type": ...  
699            NaN  {"type": "FeatureCollection", "crs": {"type": ...  

[424 rows x 6 columns]

In [128]:
main_data_merged_null_final.to_csv('Final.csv', index = False) # Saving and converting final dataframe to csv

## Data Testing - Final step
As the dictionary dataframe is ready, I need to join `Final.csv` dataset with `JOIN.csv` that has all necessary variables (volume of petrol, percentage of petrol, fuel, type of fuel) for each petrol station.

In [1]:
import pandas as pd # managing dataframes.
import numpy as np  
import math # math & numpy are for numerical/mathematical operations. 
import requests # needed to extract coordinates from physical addresses of petrol stations that lack coordinates. 
main_data_merged_null_final = pd.read_csv('Final.csv') # Our prepared dictionary dataset 
join = pd.read_csv('JOIN.csv') # Main dataset with important variables


# Currently, what we need is only two regions, so I filter them in join.csv
join = join[(join['REGION'] == 'город Алматы') | (join["REGION"] == 'город Астана')]

# Checking if everything was done correctly
print(join["REGION"].unique())
print()
print(join.info())
print()
print(main_data_merged_null_final.info())

['город Астана' 'город Алматы']

<class 'pandas.core.frame.DataFrame'>
Int64Index: 852 entries, 15 to 5271
Data columns (total 13 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   REGION             852 non-null    object 
 1   KATO               852 non-null    int64  
 2   KATO2              852 non-null    int64  
 3   ORGANIZATION       852 non-null    object 
 4   AZS                852 non-null    object 
 5   NEFTEPRODUKT_NAME  852 non-null    object 
 6   NEFTEPRODUKT_VID   852 non-null    object 
 7   DATE               852 non-null    object 
 8   WEIGHT             848 non-null    float64
 9   VOLUME             852 non-null    float64
 10  MAXVOLUME          852 non-null    float64
 11  PROC               852 non-null    float64
 12  SDU_LOAD_IN_DT     852 non-null    object 
dtypes: float64(4), int64(2), object(7)
memory usage: 93.2+ KB
None

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 424 entries, 0 to 423

In [22]:
main_data_merged_null_final

AZS  \
0                                     ТОО "МНПЗ "AMAL"   
1                                             АЗС Амал   
2                                        АЗС №12 "Ойл"   
3                                        АЗС Аурика №3   
4                                        АЗС Аурика №1   
..                                                 ...   
419                                   АЗС №129 "SOCAR"   
420  АЗС 156 ТОО "Газпромнефть-Казахствн Операционн...   
421  АЗС 157 ТОО "Газпромнефть-Казахствн Операционн...   
422                              Азс Гелиос Алматы №89   
423                                   АЗС №101 "SOCAR"   

                                               Address   Latitude  Longitude  \
0               Казахстан, г.Астана, ул. Ондирис, 60/1  51.261417  71.368687   
1          Казахстан, г.Астана, Шоссе Ондирис, д. 60/1  51.246101  71.379093   
2    Казахстан, г.Астана, ул. Тауелсыздык проспект, 4а  51.150613  71.457811   
3              Казахстан, г.Астана, ул. Жирентаева 6/1  51.148455  71.475729   
4                    Казахстан, г.Астана, пр. Туран 95  51.073720  71.388683   
..                                                 ...        ...        ...   
419            Казахстан, г.Алматы, ул.Бокейханова 241  43.299633  76.898533   
420            Казахстан, г.Алматы, ул. Ержанова 135/1  43.333905  76.937830   
421        Казахстан, г.Алматы, мкр. Казахфильм 34 "Б"  43.196480  76.905963   
422  Казахстан, г.Алматы, мкр Акжар, ул. Даулеткере...  43.181349  76.805796   
423            Казахстан, г.Алматы, мкр. Болашак д.1/7  43.285475  76.999542   

         ShortName                                     geojson_region  
0       ТОО "Амал"  {"type": "FeatureCollection", "crs": {"type": ...  
1         АЗС Амал  {"type": "FeatureCollection", "crs": {"type": ...  
2    АЗС №12 "Ойл"  {"type": "FeatureCollection", "crs": {"type": ...  
3    АЗС Аурика №3  {"type": "FeatureCollection", "crs": {"type": ...  
4        Аурика №1  {"type": "FeatureCollection", "crs": {"type": ...  
..             ...                                                ...  
419            NaN  {"type": "FeatureCollection", "crs": {"type": ...  
420            NaN  {"type": "FeatureCollection", "crs": {"type": ...  
421            NaN  {"type": "FeatureCollection", "crs": {"type": ...  
422            NaN  {"type": "FeatureCollection", "crs": {"type": ...  
423            NaN  {"type": "FeatureCollection", "crs": {"type": ...  

[424 rows x 6 columns]

In [2]:
# Joining main dataset with join data set that has necessary variables
df_joined = main_data_merged_null_final.merge(join, on = 'AZS', how = 'left')
df_joined.info()
df_joined

<class 'pandas.core.frame.DataFrame'>
Int64Index: 1023 entries, 0 to 1022
Data columns (total 18 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   AZS                1023 non-null   object 
 1   Address            1023 non-null   object 
 2   Latitude           1023 non-null   float64
 3   Longitude          1023 non-null   float64
 4   ShortName          927 non-null    object 
 5   geojson_region     1023 non-null   object 
 6   REGION             843 non-null    object 
 7   KATO               843 non-null    float64
 8   KATO2              843 non-null    float64
 9   ORGANIZATION       843 non-null    object 
 10  NEFTEPRODUKT_NAME  843 non-null    object 
 11  NEFTEPRODUKT_VID   843 non-null    object 
 12  DATE               843 non-null    object 
 13  WEIGHT             839 non-null    float64
 14  VOLUME             843 non-null    float64
 15  MAXVOLUME          843 non-null    float64
 16  PROC               843 n

AZS  \
0          ТОО "МНПЗ "AMAL"   
1                  АЗС Амал   
2             АЗС №12 "Ойл"   
3             АЗС №12 "Ойл"   
4             АЗС №12 "Ойл"   
...                     ...   
1018  Азс Гелиос Алматы №89   
1019  Азс Гелиос Алматы №89   
1020       АЗС №101 "SOCAR"   
1021       АЗС №101 "SOCAR"   
1022       АЗС №101 "SOCAR"   

                                                Address   Latitude  Longitude  \
0                Казахстан, г.Астана, ул. Ондирис, 60/1  51.261417  71.368687   
1           Казахстан, г.Астана, Шоссе Ондирис, д. 60/1  51.246101  71.379093   
2     Казахстан, г.Астана, ул. Тауелсыздык проспект, 4а  51.150613  71.457811   
3     Казахстан, г.Астана, ул. Тауелсыздык проспект, 4а  51.150613  71.457811   
4     Казахстан, г.Астана, ул. Тауелсыздык проспект, 4а  51.150613  71.457811   
...                                                 ...        ...        ...   
1018  Казахстан, г.Алматы, мкр Акжар, ул. Даулеткере...  43.181349  76.805796   
1019  Казахстан, г.Алматы, мкр Акжар, ул. Даулеткере...  43.181349  76.805796   
1020            Казахстан, г.Алматы, мкр. Болашак д.1/7  43.285475  76.999542   
1021            Казахстан, г.Алматы, мкр. Болашак д.1/7  43.285475  76.999542   
1022            Казахстан, г.Алматы, мкр. Болашак д.1/7  43.285475  76.999542   

          ShortName                                     geojson_region  \
0        ТОО "Амал"  {"type": "FeatureCollection", "crs": {"type": ...   
1          АЗС Амал  {"type": "FeatureCollection", "crs": {"type": ...   
2     АЗС №12 "Ойл"  {"type": "FeatureCollection", "crs": {"type": ...   
3     АЗС №12 "Ойл"  {"type": "FeatureCollection", "crs": {"type": ...   
4     АЗС №12 "Ойл"  {"type": "FeatureCollection", "crs": {"type": ...   
...             ...                                                ...   
1018            NaN  {"type": "FeatureCollection", "crs": {"type": ...   
1019            NaN  {"type": "FeatureCollection", "crs": {"type": ...   
1020            NaN  {"type": "FeatureCollection", "crs": {"type": ...   
1021            NaN  {"type": "FeatureCollection", "crs": {"type": ...   
1022            NaN  {"type": "FeatureCollection", "crs": {"type": ...   

            REGION          KATO  KATO2 ORGANIZATION  NEFTEPRODUKT_NAME  \
0              NaN           NaN    NaN          NaN                NaN   
1              NaN           NaN    NaN          NaN                NaN   
2     город Астана  7.100000e+09   71.0    ТОО "OIL"             Бензин   
3     город Астана  7.100000e+09   71.0    ТОО "OIL"             Бензин   
4     город Астана  7.100000e+09   71.0    ТОО "OIL"             Бензин   
...            ...           ...    ...          ...                ...   
1018  город Алматы  7.518100e+09   75.0   ТОО Гелиос  Дизельное топливо   
1019  город Алматы  7.518100e+09   75.0   ТОО Гелиос             Бензин   
1020  город Алматы  7.500000e+09   75.0  ТОО "SOCAR"             Бензин   
1021  город Алматы  7.500000e+09   75.0  ТОО "SOCAR"             Бензин   
1022  город Алматы  7.500000e+09   75.0  ТОО "SOCAR"  Дизельное топливо   

     NEFTEPRODUKT_VID                 DATE      WEIGHT    VOLUME  MAXVOLUME  \
0                 NaN                  NaN         NaN       NaN        NaN   
1                 NaN                  NaN         NaN       NaN        NaN   
2               АИ-95  2023-08-12 00:00:00  17248.0000  22728.00    26356.0   
3               АИ-92  2023-08-12 00:00:00  36344.0000  49376.00    53221.0   
4            АИ-98-K2  2023-08-12 00:00:00   4882.0000   6634.00     8898.0   
...               ...                  ...         ...       ...        ...   
1018             ДТ-Л  2023-08-12 00:00:00  17079.8535  20811.00    35235.0   
1019            АИ-95  2023-08-12 00:00:00   4686.3488   6176.00     9106.0   
1020            АИ-95  2023-08-12 00:00:00   6241.1000   8383.13    52845.4   
1021            АИ-92  2023-08-12 00:00:00  30373.3000  42202.50    60181.6   
1022             ДТ

There are only 843 rows from 1023 that were succesfully joined. I should see which rows failed to do this and drop them. These NA values exist in our dictionary dataset - `main_data_merged_null_final`. But they don't exist in database - `join.csv`

In [3]:
df_null = df_joined[df_joined['REGION'].isna()]
df_null

AZS  \
0                                      ТОО "МНПЗ "AMAL"   
1                                              АЗС Амал   
21                                       АЗС Аурика №20   
22                                       АЗС Аурика №13   
41                                         АЗС Аскар №3   
...                                                 ...   
965                           АЗС №2 ТОО "АРГА Company"   
986   АЗС 117 ТОО "Газпромнефть-Казахствн Операционн...   
999   АЗС 138 ТОО "Газпромнефть-Казахствн Операционн...   
1015  АЗС 156 ТОО "Газпромнефть-Казахствн Операционн...   
1016  АЗС 157 ТОО "Газпромнефть-Казахствн Операционн...   

                                                Address   Latitude  Longitude  \
0                Казахстан, г.Астана, ул. Ондирис, 60/1  51.261417  71.368687   
1           Казахстан, г.Астана, Шоссе Ондирис, д. 60/1  51.246101  71.379093   
21                  Казахстан, г.Астана, ул. Бекетай 14  51.182481  71.395284   
22                     Казахстан, г.Астана, ул. Озен 10  51.217664  71.381459   
41                       Казахстан, г.Астана, Манаса 12  51.158772  71.491170   
...                                                 ...        ...        ...   
965            Казахстан, г.Алматы, ул. Свободная 136/1  43.348902  76.962451   
986       Казахстан, г.Алматы, ул. Ташкентская д. 479/5  43.232573  76.824353   
999   Казахстан, г.Алматы, ул. Сагдат Нурмаганбетов,...  43.237976  76.882862   
1015            Казахстан, г.Алматы, ул. Ержанова 135/1  43.333905  76.937830   
1016        Казахстан, г.Алматы, мкр. Казахфильм 34 "Б"  43.196480  76.905963   

          ShortName                                     geojson_region REGION  \
0        ТОО "Амал"  {"type": "FeatureCollection", "crs": {"type": ...    NaN   
1          АЗС Амал  {"type": "FeatureCollection", "crs": {"type": ...    NaN   
21    АЗС Аурика 20  {"type": "FeatureCollection", "crs": {"type": ...    NaN   
22    АЗС Аурика 13  {"type": "FeatureCollection", "crs": {"type": ...    NaN   
41     АЗС Аскар №3  {"type": "FeatureCollection", "crs": {"type": ...    NaN   
...             ...                                                ...    ...   
965             NaN  {"type": "FeatureCollection", "crs": {"type": ...    NaN   
986             NaN  {"type": "FeatureCollection", "crs": {"type": ...    NaN   
999             NaN  {"type": "FeatureCollection", "crs": {"type": ...    NaN   
1015            NaN  {"type": "FeatureCollection", "crs": {"type": ...    NaN   
1016            NaN  {"type": "FeatureCollection", "crs": {"type": ...    NaN   

      KATO  KATO2 ORGANIZATION NEFTEPRODUKT_NAME NEFTEPRODUKT_VID DATE  \
0      NaN    NaN          NaN               NaN              NaN  NaN   
1      NaN    NaN          NaN               NaN              NaN  NaN   
21     NaN    NaN          NaN               NaN              NaN  NaN   
22     NaN    NaN          NaN               NaN              NaN  NaN   
41     NaN    NaN          NaN               NaN              NaN  NaN   
...    ...    ...          ...               ...              ...  ...   
965    NaN    NaN          NaN               NaN              NaN  NaN   
986    NaN    NaN          NaN               NaN              NaN  NaN   
999    NaN    NaN          NaN               NaN              NaN  NaN   
1015   NaN    NaN          NaN               NaN              NaN  NaN   
1016   NaN    NaN          NaN               NaN              NaN  NaN   

      WEIGHT  VOLUME  MAXVOLUME  PROC SDU_LOAD_IN_DT  
0        NaN     NaN        NaN   NaN            NaN  
1        NaN     NaN        NaN   NaN            NaN  
21       NaN     NaN        NaN   NaN            NaN  
22       NaN     NaN        NaN   NaN            NaN  
41       NaN     NaN        NaN   NaN            NaN  
...      ...     ...        ...   ...            ...  
965      NaN     NaN        NaN   NaN            NaN  
986      NaN     NaN        NaN   NaN            NaN  
999      NaN 

In [4]:
df_joined = df_joined.dropna(subset = 'REGION')
df_joined.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 843 entries, 2 to 1022
Data columns (total 18 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   AZS                843 non-null    object 
 1   Address            843 non-null    object 
 2   Latitude           843 non-null    float64
 3   Longitude          843 non-null    float64
 4   ShortName          761 non-null    object 
 5   geojson_region     843 non-null    object 
 6   REGION             843 non-null    object 
 7   KATO               843 non-null    float64
 8   KATO2              843 non-null    float64
 9   ORGANIZATION       843 non-null    object 
 10  NEFTEPRODUKT_NAME  843 non-null    object 
 11  NEFTEPRODUKT_VID   843 non-null    object 
 12  DATE               843 non-null    object 
 13  WEIGHT             839 non-null    float64
 14  VOLUME             843 non-null    float64
 15  MAXVOLUME          843 non-null    float64
 16  PROC               843 no

Bot also can't read some of the names of petrol stations. I am going to change it now.

In [5]:
df_joined['ORGANIZATION'] = df_joined['ORGANIZATION'].str.replace('ТОО "Газпромнефть-Казахстан Операционная Компания"', 'ТОО Газпром')
df_joined['ORGANIZATION'] = df_joined['ORGANIZATION'].str.replace('ТОО "Адалжан НС"', 'ТОО Адалжан')
df_joined['ORGANIZATION'] = df_joined['ORGANIZATION'].str.replace('ТОО АРНА', 'ТОО Арна')
df_joined['ORGANIZATION'] = df_joined['ORGANIZATION'].str.replace('ТОО "Адалжан НС"', 'ТОО Адалжан')
df_joined['ORGANIZATION'] = df_joined['ORGANIZATION'].str.replace('ТОО "АстанаНефтеСнаб"', 'ТОО "НефтеСнаб"')
df_joined['ORGANIZATION'] = df_joined['ORGANIZATION'].str.replace('ИП Турманов М.А.', 'ИП Турманов')

C:\Users\Beibarys Nyussupov\AppData\Local\Temp\ipykernel_19684\1985059301.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_joined['ORGANIZATION'] = df_joined['ORGANIZATION'].str.replace('ТОО "Газпромнефть-Казахстан Операционная Компания"', 'ТОО Газпром')
C:\Users\Beibarys Nyussupov\AppData\Local\Temp\ipykernel_19684\1985059301.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_joined['ORGANIZATION'] = df_joined['ORGANIZATION'].str.replace('ТОО "Адалжан НС"', 'ТОО Адалжан')
C:\Users\Beibarys Nyussu

In [27]:
df_joined.to_csv('Final_joined.csv', index = False)

In [39]:
df_joined["NEFTEPRODUKT_VID"].unique()

array(['АИ-95', 'АИ-92', 'АИ-98-K2', 'ДТ-З', 'ДТ-Л', 'керосин ТС-1',
       'Аи-80', 'АИ-95-К4', 'АИ-92-К5', 'Prime АИ-95', 'ДТ',
       'Вид нефтепродукта не определен', 'Prime АИ-92', 'ДТ-М-К2',
       'АИ-92-К4', 'ДТ-Л-К4', 'ДТ-Е', 'АИ-100', 'G-95', 'АИ-98-К5',
       'Prime ДТЛ', 'Аи-98-К4', 'ДТ-хр ', 'G-100'], dtype=object)

In [8]:
df_joined["ORGANIZATION"].unique()

array(['ТОО "OIL"', 'ТОО Аурика', 'ТОО Аскар Мунай Трейд',
       'ТОО "Новая АЗС"', 'ТОО Гелиос', 'ТОО "In Oil LTD KZ"',
       'ТОО "С-Мунай"', 'ТОО "Автогаз Трэйд"', 'ТОО "PetroKaz"',
       'ТОО "Sinooil"', 'ИП "Жетписова Г.К."', 'ИП "Мажитов"',
       'ТОО "Бизнес Петролеум"', 'ТОО "СБР-Астана"', 'ТОО "Royal Petrol"',
       'ИП Яковлев Ю.А.', 'ТОО "Агатай"', 'ТОО "SOCAR"', 'ПТ "AMAL"',
       'ТОО "Автосервис Лига"', 'ТОО "NurAiya Astana"',
       'ТОО "Адалжан НС"', 'ТОО АРНА', 'ТОО "АстанаНефтеСнаб"',
       'ИП Турманов М.А.',
       'ТОО "Газпромнефть-Казахстан Операционная Компания"',
       'ТОО "Munai Market"', 'ТОО "Oil Trade Center"', 'ПК "Прогресс"',
       'ТОО "Астана Мунай"', 'ИП Абитов Г.М.', 'ТОО "PETROMOBIL INVEST"',
       'ТОО "PETRO TRADING"', 'ТОО "ALMA PETROLEUM"', 'ТОО "PetroJet"'],
      dtype=object)

Our data is ready, I've added a function below that calculates the distance between coordinates given by user and coordinates of specific petrol station.

In [75]:
def haversine(lat1, lon1, lat2, lon2):
    # Convert degrees to radians
    lat1 = math.radians(lat1)
    lon1 = math.radians(lon1)
    lat2 = math.radians(lat2)
    lon2 = math.radians(lon2)
    
    # Haversine formula
    dlat = lat2 - lat1
    dlon = lon2 - lon1
    a = math.sin(dlat / 2) ** 2 + math.cos(lat1) * math.cos(lat2) * math.sin(dlon / 2) ** 2
    c = 2 * math.atan2(math.sqrt(a), math.sqrt(1 - a))
    distance = 6371.01 * c  # Earth's radius in kilometers
    
    return distance

def find_nearest_place(current_lat, current_lon, places_df, lat_column, lon_column, name_column):
    nearest_place = None
    min_distance = float('inf')
    
    for index, row in places_df.iterrows():
        place_name = row[name_column]
        place_lat = row[lat_column]
        place_lon = row[lon_column]
        distance = haversine(current_lat, current_lon, place_lat, place_lon)
        
        if distance < min_distance:
            min_distance = distance
            nearest_place = place_name
    
    return nearest_place

# Example usage
current_lat = 'WRITE_YOUR_LATITUDE'
current_lon = 'WRITE_YOUR_LANGITUDE'

# Replace this with your actual DataFrame

df_joined = df_joined[df_joined['ORGANIZATION'].str.contains('Гелиос')]
nearest_place = find_nearest_place(current_lat, current_lon, df_joined, 'Latitude', 'Longitude', 'AZS')
print("Nearest place:", nearest_place)
nearest_place = df_joined[df_joined['AZS'] == nearest_place]

TypeError: must be real number, not str